### PR TITLE
Extend Contribution Guide

### PR DESCRIPTION
**Pull request by lukaszstolarczuk**
We should consider this as the first version of this document. It should be updated with more content as we go.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+Note that if you do not provide enough information to reproduce the issue, we may not be able to take action on your report.
+Remember this is just a minimal template. You can extend it with data you think may be useful.
+-->
+
+<!-- fill the title of issue with short description -->
+
+## Environment Information
+
+- UMF version(s):                                             <!-- fill this out -->
```
Comment (igchor) [inline .github/ISSUE_TEMPLATE/bug_report.md:18]: Perhaps a 'git tag' for now since we don't really have any versions released yet?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,8 +1,74 @@
...
+
+.clang-format
+UMF_FORMAT_CODE_STYLE
+make clang-format-check & apply
+DEVELOPER_MODE
+
+**NOTE**: We use specific clang-format - version exactly **15.0** is required.
+
+<!--
+### Naming Convention
```
Comment (igchor) [inline CONTRIBUTING.md:65]: That's a good question ;d Originally we had the convention as in UR: camelCase for function names and snake_case for type names. However, I think now we also have some files where functions are named using snake_case - it would be nice to unify that.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -55,3 +121,28 @@ Use your real name (sorry, no pseudonyms or anonymous contributions.)
 
 If you set your `user.name` and `user.email` git configs, you can sign your
 commit automatically with `git commit -s`.
+
+
+<!--
+this is copy-pasted from pmemstream, it was used for OSPDT updates
```
Comment (igchor) [inline CONTRIBUTING.md:127]: I think this is a good practice so I would leave it.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,8 +1,74 @@
...
+
+- [Opening new issues](#opening-new-issues)
+- [Submitting Pull Requests](#submitting-pull-requests)
+    - [Code style](#code-style)
+    - [License](#license)
+    - [Sign your work](#sign-your-work)
+    - [Adding new dependency](#adding-new-dependency)
+    - [Extending Public API](#extending-public-api)
+
+Below you'll find instructions on how to contribute to UMF, either with code changes
```
Comment (PatKamin) [inline CONTRIBUTING.md:11]: Using either with more than two alternatives makes me unease. Maybe rephrase it a bit:
```
Below you'll find instructions on how to contribute to UMF. This can be done in three ways:
- open a PR with code changes
- file an issue with a problem to solve
- file an issue with an enhancement proposal.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,8 +1,74 @@
...
+### Building and testing
+
+described in Readme...
+plus info about tests
+
+### Code style
+
+.clang-format
+UMF_FORMAT_CODE_STYLE
+make clang-format-check & apply
```
Comment (PatKamin) [inline CONTRIBUTING.md:59]: Use portable CMake commands.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,8 +1,74 @@
...
+in the repository as a result of `Generating Source`_.
+
+Before your pull request is merged it *must* pass all jobs in the GitHub
+Actions workflow and *must* be reviewed by no less than two code owners.
+
+-->
+
+### Building and testing
+
+described in Readme...
```
Comment (PatKamin) [inline CONTRIBUTING.md:52]: Remember to add a link to README (or even the exact section of it).

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,8 +1,74 @@
...
+plus info about tests
+
+### Code style
+
+.clang-format
+UMF_FORMAT_CODE_STYLE
+make clang-format-check & apply
+DEVELOPER_MODE
+
+**NOTE**: We use specific clang-format - version exactly **15.0** is required.
```
Comment (PatKamin) [inline CONTRIBUTING.md:62]: (...) clang-format version - exactly (...)

File: CONTRIBUTING.md (hunk 2 above)
Comment (bratpiorka) [reply CONTRIBUTING.md:65]: Please remove this section for the moment - I added a JIRA UMFW-248 for changing code to use single naming convention in all files + adding a section about it here to this doc

**Review by igchor — APPROVED** [review]

**Review by bratpiorka — APPROVED** [review]

**Comment by lukaszstolarczuk** [comment]
I've rebased, made a little update in Readme file (added a link to CONTRIBUTING.md) and updated a little content of code-style and test-and-build sections.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,12 +1,100 @@
-# Contributing
+# Contributing to UMF (Unified Memory Framework)
+
+- [Opening new issues](#opening-new-issues)
+- [Submitting Pull Requests](#submitting-pull-requests)
+    - [Building and testing](#building-and-testing)
+    - [Code style](#code-style)
+    <!-- - [Naming convention](#naming-convention) -->
```
Comment (PatKamin) [inline CONTRIBUTING.md:7]: Either remove this or move to the top of this list.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,12 +1,100 @@
...
+`UMF_DEVELOPER_MODE`. To read more about all available CMake options please see
+["CMake standard options"](./README.md#cmake-standard-options) section in the top-level Readme.
+
+### Code style
+
+We use `clang-format` to verify and apply code style changes. To see all rules we require,
+please take a look at `.clang-format` file in the root directory of this repository.
+
+To enable code style checks and re-formatting, CMake option `UMF_FORMAT_CODE_STYLE` has to
+be switched on. You'll then have two additional commands available.
```
Comment (PatKamin) [inline CONTRIBUTING.md:64]: commands -> CMake targets ?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,12 +1,100 @@
...
 ### License
 
-<PROJECT NAME> is licensed under the terms in [LICENSE]<link to license file in repo>. By contributing to the project, you agree to the license and copyright terms therein and release your contribution under these terms.
+Unified Memory Framework is licensed under the terms in [LICENSE](./LICENSE.TXT). By contributing to the project,
+you agree to the license and copyright terms therein and release your contribution under these terms.
 
 ### Sign your work
 
-Please use the sign-off line at the end of the patch. Your signature certifies that you wrote the patch or otherwise have the right to pass it on as an open-source patch. The rules are pretty simple: if you can certify
+Please use the sign-off line at the end of the patch. Your signature certifies that you wrote the patch
```
Comment (PatKamin) [inline CONTRIBUTING.md:96]: I thought that signing-off wasn't meant to be a requirement in this project, @bratpiorka ?

File: CONTRIBUTING.md (hunk 8 above)
Comment (lukaszstolarczuk) [reply CONTRIBUTING.md:7]: why? it doesn't break the list and when we add this section it's going to be right about here 😉

File: CONTRIBUTING.md (hunk 9 above)
Comment (lukaszstolarczuk) [reply CONTRIBUTING.md:64]: Done.

File: CONTRIBUTING.md (hunk 10 above)
Comment (lukaszstolarczuk) [reply CONTRIBUTING.md:96]: It's still not a requirement - it's merely an ask 😉 Also, for people in our organization this may not be needed, but perhaps for outside contributors, it's a good thing to ask...?

**Review by PatKamin — APPROVED** [review]

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,12 +1,100 @@
...
+## Submitting Pull Requests
+
+We take outside code contributions to UMF through GitHub pull requests.
+You must first create your own fork of the project and submit your changes to a branch.
+You can then raise a pull request targeting `oneapi-src/unified-memory-framework:main`.
+Please try to keep commits neat and in order - please squash your commits to
+include only relevant ones (no "fixes after review" or similar).
+
+Before your pull request is merged it *must* pass all jobs in the GitHub
+Actions workflows and *must* be reviewed by no less than two reviewers.
```
Comment (bratpiorka) [inline CONTRIBUTING.md:38]: Please add a section about the review. Add info that "codeowners" would get a notification when PR is created, but one could always add specific people if needed which should speed up the review process. Also add info that PR is merged when CI passes, all discussions are resolved and PR got at least two approves.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,12 +1,100 @@
...
+be switched on. You'll then have two additional CMake targets available.
+
+To verify correct coding style of your changes execute (assuming `build` is your build directory):
+
+```bash
+$ cmake -B build -DUMF_FORMAT_CODE_STYLE=ON
+$ cmake --build build --target clang-format-check
+```
+
+If any issues were found, apply proper formatting via command:
```
Comment (bratpiorka) [inline CONTRIBUTING.md:73]: Add that the issues in formatting will fail CI and block the merge

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,12 +1,100 @@
...
+<!--
+### Naming convention
+TODO: add this section and re-format whole codebase to use such convention
+-->
 
 ### License
 
-<PROJECT NAME> is licensed under the terms in [LICENSE]<link to license file in repo>. By contributing to the project, you agree to the license and copyright terms therein and release your contribution under these terms.
+Unified Memory Framework is licensed under the terms in [LICENSE](./LICENSE.TXT). By contributing to the project,
+you agree to the license and copyright terms therein and release your contribution under these terms.
```
Comment (bratpiorka) [inline CONTRIBUTING.md:92]: Add that all new files should start from a license.

File: CONTRIBUTING.md (hunk 10 above)
Comment (bratpiorka) [reply CONTRIBUTING.md:96]: I would remove it now. I am not sure if even my approvals are signed. This is also not tested in CI. IMO this whole file contains not questions, but requirements that would fail CI if not met.

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 10 above)
Comment (bratpiorka) [reply CONTRIBUTING.md:130]: Again - could we test this? Even if not now, then in the future? If we leave it, the whole UMF team should respect it and we should not merge any non-signed-off code. But if it is a 'nice to have', I would not mention it to avoid unnecessary misunderstandings and discussions.

Comment (lukaszstolarczuk) [reply CONTRIBUTING.md:130]: heh, we had a clause like this in all(?) our previous repos, and I cannot remember a single discussion about it 😉 
We can test anything, including this (but I hope there's a tool for that 😉 ).

// examples: [PMDK](https://github.com/pmem/pmdk/blob/master/CONTRIBUTING.md) / [pmemkv](https://github.com/pmem/pmemkv/blob/master/CONTRIBUTING.md)

That said, we can use similar wording to one in PMDK, to be more flexible about it:

>In case of any doubt, the gatekeeper may ask you to certify the above in writing, i.e. via email or by including a Signed-off-by: line at the bottom of your commit comments.

Comment (bratpiorka) [reply CONTRIBUTING.md:130]: "Anecdotal evidence" :D I'm ok with "In case of any doubt" form. If we want to make it stronger we have to follow it + add tests.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,12 +1,134 @@
...
+
+# Remember to review introduced changes
+```
+
+**NOTE**: We use specific clang-format version - **15.0** is required. It can be installed,
+e.g., with command: `python -m pip install clang-format==15.0.7`.
+
+### When my PR is merged?
+
+To get your Pull Request (PR) merged you have to fulfill several requirements.
```
Comment (bratpiorka) [inline CONTRIBUTING.md:87]: I'm not an expert but I think this could sound better:

To get your Pull Request (PR) to be merged, you must meet several requirements - the basic are:
- The project compiles correctly,
- ....

Comment (lukaszstolarczuk) [reply CONTRIBUTING.md:87]: This way is less fun 😄😉 Rewritten.

File: CONTRIBUTING.md (hunk 10 above)
Comment (lukaszstolarczuk) [reply CONTRIBUTING.md:130]: Rewritten.

**Review by bratpiorka — APPROVED** [review]

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,13 +1,129 @@
...
 
 ### License
 
-<PROJECT NAME> is licensed under the terms in [LICENSE]<link to license file in repo>. By contributing to the project, you agree to the license and copyright terms therein and release your contribution under these terms.
+Unified Memory Framework is licensed under the terms in [LICENSE](./LICENSE.TXT) file. By contributing to the project,
+you agree to the license and copyright terms therein and release your contribution under these terms.
 
-### Sign your work
+**Note:**
+>Each new file added to the repository has to contain the appropriate license header. To see how such header
```
Comment (PatKamin) [inline CONTRIBUTING.md:122]: such (a?) header **looks** like

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -1,13 +1,129 @@
...
+Your Pull Request (PR) will be merged if you meet several requirements - the basic are:
+- The project builds properly,
+- All tests are executed, and no issues have been reported,
+- All checks pass (code style, spelling, etc.),
+- Additional requirements are fulfilled (e.g., see below, for a new public API function).
+
+While the most of these requirements are verified via automated scripts run in
+Continuous Integration (CI) they are also verified with human touch - **the review**!
+
+Code review has to be done by at least two UMF maintainers. They "maintainers team"
```
Comment (PatKamin) [inline CONTRIBUTING.md:95]: They -> The

Comment (lukaszstolarczuk) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md (hunk 15 above)
Comment (lukaszstolarczuk) [reply CONTRIBUTING.md]: Done.

**Review by PatKamin — APPROVED** [review]
